rfcs/0000-structured-logging.md

```diff
@@ -0,0 +1,2189 @@
+
+`log` is a logging _facade_. It's only concerned with a standard, minimal API for capturing log records, and surfacing those records to some consumer. The tools provided by `log` are only those that are fundamental to the operation of the `log!` macro. From `log`'s point of view, a logging framework like `slog` is a black-box implementation of the `Log` trait. In this role, the `Log` trait can act as a common entrypoint for capturing log records. That means the `Record` type can act as a common container for describing a log record. `log` has its own set of trade-offs baked into the design of its API. The `log!` macro assumes a single, global entrypoint, and all data in a log record is borrowed from the callsite.
+
+A healthy logging ecosystem needs both `log` and frameworks like `slog`. As a standard API, `log` can support a diverse but cohesive ecosystem of logging tools in Rust by acting as the glue between libraries, frameworks, and applications. A lot of libraries already depend on it. In order to really fulfil this role though, `log` needs to support structured logging so that libraries and their consumers can take advantage of it in a framework-agnostic way.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Logging structured key-value pairs
+
+Structured logging is supported in `log` by allowing typed key-value pairs to be associated with a log record. A `;` separates structured key-value pairs from values that are replaced into the message:
+
+```rust
+info!(
+    "This is the rendered {message}. It is not structured",
+    message = "message";
+    correlation = correlation_id,
+    user
+);
+```
+
+Any `value` or `key = value` expressions before the `;` in the macro will be interpolated into the message as unstructured text. This is the `log!` macro we have today. Any `value` or `key = value` expressions after the `;` will be captured as structured key-value pairs. These structured key-value pairs can be inspected or serialized, retaining some notion of their original type. That means in the above example, the `message` key is unstructured, and the `correlation` and `user` keys are structured:
+
+```
+info!(
+    "This is the rendered {message}. It is not structured",
+    message = "message";
+    ^^^^^^^^^^^^^^^^^^^
+    unstructured
+
+    correlation = correlation_id,
+    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+    structured
+
+    user
+    ^^^^
+    structured
+);
+```
+
+### What can be logged?
+
+Anything that's `std::fmt::Debug + serde::Serialize` can be logged as a structured value. Using default crate features though, only a fixed set of types from the standard library are supported:
+
+- Standard formats: `Arguments`
+- Primitives: `bool`, `char`
+- Unsigned integers: `u8`, `u16`, `u32`, `u64`, `u128`
+- Signed integers: `i8`, `i16`, `i32`, `i64`, `i128`
+- Strings: `&str`, `String`
+- Bytes: `&[u8]`, `Vec<u8>`
+- Paths: `Path`, `PathBuf`
+- Special types: `Option<T>` and `()`
```

Comment:
Maybe support types that `Deref`/`AsRef`/`Borrow` into one of these types as well? For example I have a couple of that are defined as `struct MyId(usize)`, would be nice if those were supported out of the box if it implemets one of those traits.

But this can wait until later perhaps.

Reply:
Unfortunately something like `impl<T: AsRef<U>, U: Visit>` wouldn't be coherent, `impl<T: Deref<Target = U>, U: Visit>` would only cover smart pointers, and as a blanket implementation it would conflict with the optional one for `T: Serialize + Debug`. This would be a case where if you want to log `MyId(usize)`, you can enable the `kv_serde` feature and derive `Serialize` on `MyId` to log it as a newtype.